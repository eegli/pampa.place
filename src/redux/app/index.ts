import {PaletteMode} from '@mui/material';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';

interface AppState {
  authError: boolean;
  apiKey: string | undefined;
  theme: PaletteMode;
}

const initialState: AppState = {
  authError: false,
  apiKey: undefined,
  theme: 'dark',
};

const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers: {
    setApiKey(state, {payload}: PayloadAction<string | undefined>) {
      state.apiKey = payload;
    },
    setTheme(state, {payload}: PayloadAction<PaletteMode>) {
      state.theme = payload;
    },
  },
});

export const {setApiKey, setTheme} = appSlice.actions;
export default appSlice;
