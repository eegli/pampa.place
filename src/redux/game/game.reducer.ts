import config, { LatLngLiteral, MapData } from '@config';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { calcDist, calcScore, OrNull } from '@utils';

export type Result = {
  round: number;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

type Player = {
  totalScore: number;
} & { results: Result[] };

export enum STATUS {
  PENDING = 'PENDING',
  ROUND_STARTED = 'ROUND_STARTED',
  INTERMISSION = 'INTERMISSION',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

interface GameState {
  initialized: boolean;
  map: MapData;
  players: string[];
  playerInfo: Record<string, Player>;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
  status: STATUS;
}

const initialState: GameState = {
  initialized: false,
  map: config.maps[0],
  players: [],
  playerInfo: {},
  rounds: {
    current: 1,
    total: 3,
    progress: 0,
  },
  status: STATUS.PENDING,
};

const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    initGame(
      state,
      action: PayloadAction<{
        names: string[];
        rounds: number;
        map: MapData;
      }>
    ) {
      state.players = action.payload.names;

      action.payload.names.forEach(name => {
        state.playerInfo[name] = { totalScore: 0, results: [] };
      });

      state.map = action.payload.map;

      state.rounds.total = action.payload.rounds;
      state.rounds.current = 1;
      state.rounds.progress = 0;

      state.status = STATUS.INTERMISSION;
      state.initialized = true;
    },
    finishRound(state) {
      state.rounds.current++;
      state.rounds.progress = 0;

      // Game has ended
      if (state.rounds.total < state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        state.status = STATUS.INTERMISSION;
      }
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Set current round score
      const player = state.playerInfo[state.players[0]];

      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(state.map.computed.area, dist);
        selected = payload.selected;
      }

      const newScore = player.totalScore + score;
      const existingResults = player.results;

      state.playerInfo[state.players[0]] = {
        totalScore: newScore,
        results: [
          ...existingResults,
          {
            round: state.rounds.current,
            selected,
            dist,
            score,
          },
        ],
      };

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.pop();
      if (last) state.players.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.length) {
        // Display popup
        state.status = STATUS.INTERMISSION;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
  },
});

export const { reset, initGame, startRound, finishRound, setPlayerScore } =
  gameSlice.actions;
export default gameSlice.reducer;
