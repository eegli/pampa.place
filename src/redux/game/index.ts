import {config} from '@/config/game';
import {MAPS, MAP_IDS} from '@/config/maps';
import {LatLngLiteral} from '@/config/types';
import {calcDist, calcScore} from '@/utils/geo';
import {OrNull} from '@/utils/types';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';

export type Result = {
  name: string;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

export enum STATUS {
  UNINITIALIZED = 'UNINITIALIZED',
  ROUND_STARTED = 'ROUND_STARTED',
  PENDING_PLAYER = 'PENDING_PLAYER',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

export interface GameState {
  status: STATUS;
  mapId: string;
  mapName: string;
  players: string[];
  // Store the results for each round in the outer array. The innner
  // array consists of player result objects.
  scores: Result[][];
  timeLimit: number;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
}

const initialState: GameState = {
  status: STATUS.UNINITIALIZED,
  mapId: MAP_IDS[0].id,
  mapName: MAP_IDS[0].name,
  players: [],
  scores: [],
  timeLimit: config.timeLimitsDefault,
  rounds: {
    current: 1,
    total: config.roundsDefault,
    progress: 0,
  },
};

const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    setPlayers(state, action: PayloadAction<string[]>) {
      state.players = action.payload.filter(Boolean).map(el => el.slice(0, 24));
    },
    setRounds(state, action: PayloadAction<number>) {
      state.rounds.total = action.payload;
    },
    setTimeLimit(state, action: PayloadAction<number>) {
      state.timeLimit = action.payload;
    },
    setMap(state, action: PayloadAction<string>) {
      state.mapId = action.payload;
      state.mapName = MAPS[action.payload].feature.properties.name;
    },
    initGame(state) {
      if (!state.players.length) {
        state.players = ['Player 1'];
      }
      state.scores = [];
      // Empty score array for first round
      state.scores.push([]);
      state.rounds.current = 1;
      state.rounds.progress = 0;
      state.status = STATUS.PENDING_PLAYER;
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(MAPS[state.mapId].area, dist);
        selected = payload.selected;
      }

      state.scores[state.rounds.current - 1].push({
        name: state.players[0],
        selected,
        dist,
        score,
      });

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.pop();
      if (last) state.players.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.length) {
        // Display popup
        state.status = STATUS.PENDING_PLAYER;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
    resetRound(state) {
      // Reset score for this round
      state.scores[state.rounds.current - 1] = [];

      // Reset round progress
      state.rounds.progress = 0;

      // Rotate to first player again
      for (let i = 0; i < state.rounds.progress; i++) {
        const playerName = state.players.pop();
        if (playerName) {
          state.players.unshift(playerName);
        }
      }

      // Set zero for current round scores
      state.status = STATUS.PENDING_PLAYER;
    },
    endRound(state) {
      // Game has ended
      if (state.rounds.total === state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        // Prepare for next round
        state.rounds.current++;
        state.rounds.progress = 0;
        state.scores.push([]);
        state.status = STATUS.PENDING_PLAYER;
      }
    },
  },
});

export const {
  reset,
  initGame,
  startRound,
  endRound,
  resetRound,
  setPlayerScore,
  setRounds,
  setMap,
  setPlayers,
  setTimeLimit,
} = gameSlice.actions;
export default gameSlice;
