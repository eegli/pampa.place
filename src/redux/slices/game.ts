import { config } from '@/config/game';
import { CUSTOM_MAP_IDS, LatLngLiteral, MAPS } from '@/config/maps';
import { calcDist, calcScore } from '@/utils/geo';
import { OrNull } from '@/utils/types';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export type Result = {
  round: number;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

type Player = {
  totalScore: number;
} & { results: Result[] };

export enum STATUS {
  PENDING_START = 'PENDING_START', // Uninitialized
  ROUND_STARTED = 'ROUND_STARTED',
  INTERMISSION = 'INTERMISSION',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

export interface GameState {
  status: STATUS;
  mapId: string;
  mapName: string;
  players: {
    names: string[];
    scores: Record<string, Player>;
  };
  timeLimit: number;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
}

export const initialState: GameState = {
  status: STATUS.PENDING_START,
  mapId: CUSTOM_MAP_IDS[0].id,
  mapName: CUSTOM_MAP_IDS[0].name,
  players: {
    names: [],
    scores: {},
  },
  timeLimit: config.timeLimitsDefault,
  rounds: {
    current: 1,
    total: config.roundsDefault,
    progress: 0,
  },
};

export const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    setPlayers(state, action: PayloadAction<string[]>) {
      state.players.names = action.payload
        .filter(Boolean)
        .map(el => el.slice(0, 24));
    },
    setRounds(state, action: PayloadAction<number>) {
      state.rounds.total = action.payload;
    },
    setTimeLimit(state, action: PayloadAction<number>) {
      state.timeLimit = action.payload;
    },
    setMap(state, action: PayloadAction<string>) {
      state.mapId = action.payload;
      state.mapName = MAPS[action.payload].geo.properties.name;
    },
    initGame(state) {
      if (!state.players.names.length) {
        state.players.names = ['Player 1'];
      }

      state.players.scores = {};

      state.players.names.forEach(name => {
        state.players.scores[name] = { totalScore: 0, results: [] };
      });

      state.rounds.current = 1;
      state.rounds.progress = 0;

      state.status = STATUS.INTERMISSION;
    },
    finishRound(state) {
      state.rounds.current++;
      state.rounds.progress = 0;

      // Game has ended
      if (state.rounds.total < state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        state.status = STATUS.INTERMISSION;
      }
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    resetRound(state) {
      // Rotate to first player again
      for (let i = 0; i < state.rounds.progress; i++) {
        const playerName = state.players.names.pop();
        if (playerName) {
          // Remove score from this round
          const result = state.players.scores[playerName].results.pop();

          // Subtract score from this round from total score
          if (result)
            state.players.scores[playerName].totalScore -= result.score;

          // Rotate player
          state.players.names.unshift(playerName);
        }
      }

      // Reset round progress
      state.rounds.progress = 0;

      // Set zero for current round scores
      state.status = STATUS.INTERMISSION;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Set current round score
      const player = state.players.scores[state.players.names[0]];

      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(MAPS[state.mapId].computed.area, dist);
        selected = payload.selected;
      }

      const newScore = player.totalScore + score;
      const existingResults = player.results;
      state.players.scores[state.players.names[0]] = {
        totalScore: newScore,
        results: [
          ...existingResults,
          {
            round: state.rounds.current,
            selected,
            dist,
            score,
          },
        ],
      };

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.names.pop();
      if (last) state.players.names.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.names.length) {
        // Display popup
        state.status = STATUS.INTERMISSION;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
  },
});

export const {
  reset,
  initGame,
  startRound,
  finishRound,
  resetRound,
  setPlayerScore,
  setRounds,
  setMap,
  setPlayers,
  setTimeLimit,
} = gameSlice.actions;
export default gameSlice.reducer;
