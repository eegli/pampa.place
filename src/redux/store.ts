import {Constants} from '@/config/constants';
import {
  configureStore,
  Middleware,
  StateFromReducersMapObject,
} from '@reduxjs/toolkit';
import {createLogger} from 'redux-logger';
import app from './app';
import game from './game';
import position from './position';

const reducer = {
  position: position.reducer,
  app: app.reducer,
  game: game.reducer,
};

const logger = createLogger();

const windowStorage: Middleware<{}, RootState> = state => next => action => {
  if (action.type.includes('setApiKey') && typeof action.payload === 'string') {
    window.sessionStorage.setItem(Constants.SESSION_APIKEY_KEY, action.payload);
  }
  if (action.type.includes('setTheme')) {
    window.localStorage.setItem(Constants.THEME_KEY, action.payload);
  }
  return next(action);
};

const isDev = process.env.NODE_ENV === 'development';
const devMiddleware = [logger, windowStorage];
const prodMiddleware = [windowStorage];

export const initialStates = {
  position: position.reducer(undefined, {type: '@@INIT'}),
  app: app.reducer(undefined, {type: '@@INIT'}),
  game: game.reducer(undefined, {type: '@@INIT'}),
};

export function createStore(preloadedState?: RootState) {
  return configureStore({
    reducer,
    preloadedState,
    devTools: false,
    middleware: getDefault =>
      isDev
        ? getDefault().concat(devMiddleware)
        : getDefault().concat(prodMiddleware),
  });
}

export const store = createStore();

export type AppDispatch = typeof store.dispatch;
export type RootState = StateFromReducersMapObject<typeof reducer>;
