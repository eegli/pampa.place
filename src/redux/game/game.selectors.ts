import { createSelector } from '@reduxjs/toolkit';
import { RootState } from '../redux.store';

export const getActivePlayer = (s: RootState) => s.game.players[0];

export const getPlayerCount = (s: RootState) => s.game.players.length;

// TODO use memoization once redux toolkit exports reselect 4.1
export const getRoundScores = createSelector(
  [
    (s: RootState) => s.game.scores,
    (s: RootState) => s.game.rounds.current - 1,
  ],
  (scores, currentRound) => {
    /* Sort on a copy, otherwise, Redux isn't happy: https://stackoverflow.com/a/66721870 */
    return [...scores[currentRound]].sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

export const getTotalScores = createSelector(
  [(s: RootState) => s.game.scores, (s: RootState) => s.game.players],
  (scores, players) => {
    const total = Array.from({ length: players.length }, (_, idx) => ({
      name: players[idx],
      score: 0,
    }));

    scores.forEach(round => {
      round.forEach((playerResult, playerIdx) => {
        total[playerIdx].score += playerResult.score;
      });
    });

    return total.sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

export const shouldRequestNewSV = (s: RootState) =>
  s.game.rounds.progress === 0;

export const isFinished = (s: RootState) =>
  s.game.rounds.progress === s.game.players.length &&
  s.game.rounds.current === s.game.rounds.total;
