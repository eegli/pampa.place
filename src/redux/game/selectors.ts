import {createSelector} from '@reduxjs/toolkit';
import {RootState} from '../store';

export const getActivePlayer = (s: RootState) => s.game.players[0];

export const getPlayerCount = (s: RootState) => s.game.players.length;

function sortScores<T extends {score: number}[]>(scores: T): T {
  return scores.sort((a, b) =>
    a.score > b.score ? -1 : b.score > a.score ? 1 : 0
  );
}

export const getCurrentRoundScores = createSelector(
  (s: RootState) => s.game.scores,
  scores => {
    const lastScore = scores.length - 1;
    return sortScores(Array.from(scores[lastScore]));
  }
);

export const getTotalScores = createSelector(
  (s: RootState) => s.game.scores,
  (s: RootState) => s.game.players,
  (scores, players) => {
    const total = Array.from({length: players.length}, (_, idx) => ({
      name: players[idx],
      score: 0,
    }));

    scores.forEach(round => {
      round.forEach((playerResult, playerIdx) => {
        total[playerIdx].score += playerResult.score;
      });
    });

    return sortScores(total);
  }
);

export const isFinished = (s: RootState) =>
  s.game.rounds.progress === s.game.players.length &&
  s.game.rounds.current === s.game.rounds.total;
