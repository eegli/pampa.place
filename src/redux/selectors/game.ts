import { createSelector } from '@reduxjs/toolkit';
import { RootState } from '../store';

export const getActivePlayer = (s: RootState) => s.game.players[0];

export const getPlayerCount = (s: RootState) => s.game.players.length;

// TODO use memoization once redux toolkit exports reselect 4.1
export const getRoundScores = createSelector(
  [
    (s: RootState) => s.game.scores,
    (s: RootState) => s.game.rounds.current - 1,
  ],
  (scores, currentRound) => {
    /* Sort on a copy, otherwise, Redux isn't happy: https://stackoverflow.com/a/66721870 */
    return [...scores[currentRound]].sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

export const getTotalScores = createSelector(
  [(s: RootState) => s.game.scores, (s: RootState) => s.game.players],
  (scores, players) => {
    const final = Array.from({ length: players.length }, (_, idx) => ({
      name: players[idx],
      score: 0,
    }));

    scores.forEach(round => {
      round.forEach((playerResult, playerIdx) => {
        final[playerIdx].score += playerResult.score;
      });
    });

    return final.sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

const s = [
  [
    { name: 'e', round: 1, selected: 'x', dist: 1, score: 999 },
    { name: 'x', round: 1, selected: 'x', dist: 1, score: 999 },
  ],
  [
    { name: 'e', round: 2, selected: 'x', dist: 1, score: 999 },
    { name: 'x', round: 2, selected: 'x', dist: 1, score: 999 },
  ],
];

export const shouldRequestNewSV = (s: RootState) =>
  s.game.rounds.progress === 0;

export const isFinished = (s: RootState) =>
  s.game.rounds.progress === s.game.players.length &&
  s.game.rounds.current === s.game.rounds.total;
