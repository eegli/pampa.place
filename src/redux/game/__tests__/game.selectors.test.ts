import {RootState, store} from '../../redux.store';
import * as selectors from '../game.selectors';

describe('Redux, game selectors', () => {
  it('calculates round scores', () => {
    const state: RootState = {
      ...store.getState(),
      game: {
        ...store.getState().game,
        players: ['Player 1', 'Player 2'],
        scores: [
          [
            {name: 'Player 1', selected: null, dist: 2, score: 10},
            {name: 'Player 2', selected: null, dist: 2, score: 5},
          ],
          [
            {name: 'Player 1', selected: null, dist: 2, score: 20},
            {name: 'Player 2', selected: null, dist: 2, score: 10},
          ],
        ],
        rounds: {
          total: 2,
          current: 1,
          progress: 2,
        },
      },
    };

    expect(selectors.getRoundScores(state)).toMatchSnapshot();
  });
});
