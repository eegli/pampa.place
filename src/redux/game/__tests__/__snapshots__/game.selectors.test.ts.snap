// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux, game selectors calculates round scores 1`] = `
Array [
  Object {
    "dist": 2,
    "name": "Player 1",
    "score": 10,
    "selected": null,
  },
  Object {
    "dist": 2,
    "name": "Player 2",
    "score": 5,
    "selected": null,
  },
]
`;
